XMonad/Util/ClickableWorkspaces.hs

```diff
@@ -18,44 +18,72 @@ module XMonad.Util.ClickableWorkspaces (
 import XMonad.Hooks.DynamicLog (xmobarAction, xmobarRaw, PP(..))
+import XMonad.Layout.IndependentScreens
+import XMonad.Util.WorkspaceCompare (getWsIndex)
 
 -- $usage
 -- However you have set up your PP, apply @clickablePP@ to it, and bind the result
 -- to "XMonad.Hooks.DynamicLog"\'s dynamicLogWithPP like so:
--- 
+--
 -- > logHook = clickablePP xmobarPP { ... } >>= dynamicLogWithPP
 --
 -- * Requirements:
---   * wmctrl on system (in path)
---   * "XMonad.Hooks.EwmhDesktops" for wmctrl support (see Hackage docs for setup)
---   * use of UnsafeStdinReader in xmobarrc (rather than StdinReader)
+--   * @xdotool@ on system (in path)
+--   * "XMonad.Hooks.EwmhDesktops" for @xdotool@ support (see Hackage docs for setup)
+--   * use of UnsafeStdinReader/UnsafeXMonadLog in xmobarrc (rather than StdinReader/XMonadLog)
 
 
 clickableWrap :: Int -> String -> String
-clickableWrap i ws = xmobarAction ("wmctrl -s " ++ show i) "1" $ xmobarRaw ws
+clickableWrap i ws = xmobarAction ("xdotool set_desktop " ++ show i) "1" $ xmobarRaw ws
 
--- Use index of workspace in users config to target workspace with wmctrl switch.
-getClickable :: X (WorkspaceId -> String)
-getClickable = do
+-- | Use index of workspace in users config to target workspace with @xdotool@ switch.
+getClickable :: (WorkspaceId -> String) -> X (WorkspaceId -> String)
+getClickable ren = do
   wsIndex <- getWsIndex
   return $ \ws -> case wsIndex ws of
-                    Just idx -> clickableWrap idx ws
+                    Just idx -> clickableWrap idx (ren ws)
                     Nothing -> ws
 
 -- | Apply clickable wrapping to all workspace fields in given PP.
 clickablePP :: PP -> X PP
-clickablePP pp = do
-  clickable <- getClickable
+clickablePP = clickableRenamedPP id
+
+-- | Alternative to 'clickablePP' that allows changing the visible workspace
+-- name. Useful for integration with modules that change workspace names, such
+-- as "XMonad.Layout.IndependentScreens" and "XMonad.Actions.WorkspaceNames".
+-- See "XMonad.Util.ClickableWorkspaces.Integrations".
+clickableRenamedPP :: (WorkspaceId -> String) -> PP -> X PP
+clickableRenamedPP ren pp = do
+  clickable <- getClickable ren
   return $
     pp { ppCurrent         = ppCurrent pp . clickable
        , ppVisible         = ppVisible pp . clickable
        , ppHidden          = ppHidden pp . clickable
        , ppHiddenNoWindows = ppHiddenNoWindows pp . clickable
+       , ppVisibleNoWindows= ppVisibleNoWindows pp <&> (. clickable)
        , ppUrgent          = ppUrgent pp . clickable
        }
+
+-- | Integration with "XMonad.Actions.WorkspaceNames".
+clickableWorkspaceNamesPP :: PP -> X PP
+clickableWorkspaceNamesPP pp = do
+    rename <- getWorkspaceNames
+    clickableRenamedPP rename pp
+
+-- | Integration with "XMonad.Layout.IndependentScreens".
+clickableMarshallPP :: ScreenId -> PP -> X PP
+clickableMarshallPP s pp =
+    clickableRenamedPP (snd . unmarshall) pp{ ppSort = marshallSort s <$> ppSort pp }
```

Comment:
This seems to work as far as hiding the ScreenId prefixes (e.g. `0_0` and `0_1` show as `0` and `1`). I can't seem to get it to play nicely with `clickableWorkspaceNamesPP` though -- workspaces aren't displaying their names, just their numbers.

I have an admittedly complicated use-case, integrating IndependentScreens, WorkspaceNames, and DynamicBars, so I'm not opposed to providing this helper as is, but I'm curious whether it's possible to support use-cases like mine (and whether it already can but I'm just misusing it).

Here's a snippet from my logHook, when using these integration helpers. The workspaces are clickable and ScreenId prefixes are hidden, but as I mentioned, workspace names aren't shown:
```haskell
logString :: PP -> ScreenId -> X String
logString pp =
  composePP pp >=> dynamicLogString

composePP :: PP -> ScreenId -> X PP
composePP pp s = do
  pp'  <- clickableWorkspaceNamesPP pp
  pp'' <- clickableMarshallPP s pp'
  pure (filterOutWsPP [scratchpadWorkspaceTag] pp'')
```
In case it's useful, a larger snippet offering more context is here: https://gist.github.com/ivanbrennan/f1e75efe0d8c8c969f256b53a93e3510

Reply:
Oh, I totally forgot that you're using all three (clickable, independent, names)! This will need a functions that combines it all together, as these unfortunately don't compose nicely as they all need to see the original unmangled workspace tag. I'll push an update later today. Thanks for your feedback so far.

Reply:
(It would all be much better if we had a PP where the ppCurrent, ppVisible, … functions were given the original workspace (e.g. `WorkspaceId → String → String`), then they'd compose better. But I'm not sure we can change the type now. :-/)

Reply:
I think I finally figured out a nice way to make the composition of `clickablePP`, `workspaceNamesPP` and `marshallPP` just work: https://github.com/xmonad/xmonad-contrib/pull/481